### ISSUE TITLE
Filter / Directives for services

### ISSUE THREAD

**Issue by sebastian-alfers** [akka/akka-grpc#1429]
### Short description

It would be great to be able to add Filters / Directives for a given service.

### Details

If you have common logic (e.g. authentication, validation, enhancement) of an incoming grpc request if would be great if that logic can be abstracted for a given type of Request / Response. This enhances composability, reusability and helps to follow separation of concern.

**Comment by raboof** [comment]
Did you see https://doc.akka.io/docs/akka-grpc/current/server/akka-http.html? Would that help?

**Comment by sebastian-alfers** [comment]
@raboof That is a great help but using this approach I would assume that the payload was not yet parsed into the expected protobuf-generated class. Is that correct or am I missing anything?

**Comment by raboof** [comment]
You're correct that a limitation of the directive-based approach is that, in the directive, you don't have access to the marshalled request/response objects, only the 'unparsed' HTTP request/response.

For the requests, you can channel information from the directive to the 'typed' implementation via the Metadata in the power API - in the future it might be good to also support Akka HTTP Attributes there.

Possibly introducing a second kind of directives at the 'Akka gRPC level' might provide more power, but depending on what kind of approach we can come up with we'll have to balance whether that weighs up to the complexity it might add.
